### PR TITLE
Refactor robot-simulator exercise

### PR DESCRIPTION
**Pull request by artamonovkirill**
- verify results properties through getters
- make code groovier (use def key-word, remove semi-colons)
- use Spock's given/when/then
- visually separate parts of data-driven test inputs
- document Groovy's values and properties

### CODE REVIEW

File: exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy [hunk 1]
```diff
@@ -1,30 +1,51 @@
...
+    private int pos_x
+    private int pos_y
+    private String direction
 
     static allowed_directions = ['north', 'east', 'south', 'west']
 
+    /**
+     * Returns the current position of the robot relative to the X axis
+     */
+    def getX() {
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy:11]: given `getX` method is defined, groovy uses it in the code like `robot.x`

File: exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy [hunk 2]
```diff
@@ -1,30 +1,51 @@
...
+     * Returns the current position of the robot relative to the Y axis
+     */
+    int getY() {
+        return pos_y
+    }
+
+    /**
+     * Returns the direction that the robot is currently facing
+     */
+    String getDirection() {
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy:25]: By using getters, we let internal implementation use any other type to store direction (I opted for int in my solution)

File: exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy [hunk 3]
```diff
@@ -1,30 +1,51 @@
...
+        return direction
+    }
+
     RobotSimulator(int pos_x, int pos_y, String direction) {
-        this.pos_x = pos_x;
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/.meta/src/reference/groovy/RobotSimulator.groovy:9]: semi-colon are not required in Groovy

File: exercises/practice/robot-simulator/src/test/groovy/RobotSimulatorSpec.groovy [hunk 4]
```diff
@@ -1,85 +1,110 @@
...
 
 class RobotSimulatorSpec extends Specification {
 
     @Unroll
     def "Create robot at origin #label"() {
-        expect:
-        RobotSimulator results = new RobotSimulator(x, y, direction)
-        RobotSimulator expected = new RobotSimulator(x, y, direction)
-        results.properties == expected.properties
+        when:
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/src/test/groovy/RobotSimulatorSpec.groovy:9]: given/when/then to have more readable tests

File: exercises/practice/robot-simulator/src/test/groovy/RobotSimulatorSpec.groovy [hunk 5]
```diff
@@ -1,85 +1,110 @@
...
+        then:
+        results.x == x
+        results.y == y
+        results.direction == direction
 
         where:
-        label                               || x    || y   || direction
-        "at origin facing north"            || 0    || 0   || "north"
-        "at negative position facing south" || -1   || -1  || "south"
+        label                               || x  | y  | direction
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/src/test/groovy/RobotSimulatorSpec.groovy:18]: | works as a separator as well, left || only to separate distinct parts of data-driven tests input

File: exercises/practice/robot-simulator/src/main/groovy/RobotSimulator.groovy [hunk 6]
```diff
@@ -3,6 +3,27 @@ class RobotSimulator {
...
+     * Returns the current position of the robot relative to the Y axis
+     */
+    int getY() {
+        throw new UnsupportedOperationException('getY implementation is missing')
+    }
+
+    /**
+     * Returns the direction that the robot is currently facing
+     */
+    String getDirection() {
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/src/main/groovy/RobotSimulator.groovy:23]: By using getters, we let internal implementation use any other type to store direction (I opted for int in my solution)

File: exercises/practice/robot-simulator/src/main/groovy/RobotSimulator.groovy [hunk 7]
```diff
@@ -3,6 +3,27 @@ class RobotSimulator {
         throw new UnsupportedOperationException('Constructor implementation is missing')
     }
 
+    /**
+     * Returns the current position of the robot relative to the X axis
+     */
+    int getX() {
```
Comment (artamonovkirill) [inline exercises/practice/robot-simulator/src/main/groovy/RobotSimulator.groovy:9]: If `getX()` is defined, Groovy will use the getter in code like `results.x`.
I'm fine with defining simply `x()`, `y()` and `direction()` getters to minimize Groovy features in the context of the exercise

**Review by glennj — CHANGES_REQUESTED** [review]
I'm in favour of all of this. But a quick bit of instruction to new users is wanted here.

**Comment by artamonovkirill** [comment]
> I'm in favour of all of this. But a quick bit of instruction to new users is wanted here.

Added a note here: https://github.com/artamonovkirill/groovy/pull/1/commits/3a8737ee09c376e1ad545d3f274ec61a471b8ad8#diff-8639b57aa8e48068d53e221629cc931542d14f2b0e464bfd429316d722c0806f
Would that be sufficient or should I also go into explaining what getters and setters are?

**Comment by glennj** [comment]
> Added a note here: [3a8737e](https://github.com/artamonovkirill/groovy/commit/3a8737ee09c376e1ad545d3f274ec61a471b8ad8) Would that be sufficient or should I also go into explaining what getters and setters are?

This is great.

The new text must be added to a file named `instructions.append.md` -- if someone were to re-sync the exercise with the problem-specifications repo, the new text would be lost.

The Exercism markdown standards require that every markdown file has exactly one level-one heading, even if that header will be ignored. `instructions.append.md` should start like

```markdown
# ignore

## Properties and fields in Groovy

In this exercise, ...
```

**Comment by artamonovkirill** [comment]
> The new text must be added to a file named `instructions.append.md` -- if someone were to re-sync the exercise with the problem-specifications repo, the new text would be lost.

Done. As the PR seems to be reaching it's final form, I've also already created one towards the Exercism's repo: https://github.com/exercism/groovy/pull/461

**Pull request closed without merging**